apptest/tests/dedup_test.go

```diff
@@ -0,0 +1,218 @@
+package tests
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/apptest"
+	at "github.com/VictoriaMetrics/VictoriaMetrics/apptest"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/decimal"
+	pb "github.com/VictoriaMetrics/VictoriaMetrics/lib/prompbmarshal"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+func TestSingleDeduplication_dedulicationIsOff(t *testing.T) {
+	tc := at.NewTestCase(t)
+	defer tc.Stop()
+
+	sut := tc.MustStartVmsingle("vmsingle", []string{
+		"-storageDataPath=" + tc.Dir() + "/vmsingle",
+		"-retentionPeriod=100y",
+		"-dedup.minScrapeInterval=0",
+	})
+
+	testDeduplication(tc, sut, false)
+}
+
+func TestSingleDeduplication_dedulicationIsOn(t *testing.T) {
+	tc := at.NewTestCase(t)
+	defer tc.Stop()
+
+	sut := tc.MustStartVmsingle("vmsingle", []string{
+		"-storageDataPath=" + tc.Dir() + "/vmsingle",
+		"-retentionPeriod=100y",
+		"-dedup.minScrapeInterval=10s",
+	})
+
+	testDeduplication(tc, sut, true)
+}
+
+func TestClusterDeduplication_deduplicationIsOff(t *testing.T) {
+	tc := at.NewTestCase(t)
+	defer tc.Stop()
+
+	sut := tc.MustStartCluster(&at.ClusterOptions{
+		Vmstorage1Instance: "vmstorage1",
+		Vmstorage1Flags: []string{
+			"-dedup.minScrapeInterval=0",
+		},
+		Vmstorage2Instance: "vmstorage2",
+		Vmstorage2Flags: []string{
+			"-dedup.minScrapeInterval=0",
+		},
+		VminsertInstance: "vminsert",
+		VmselectInstance: "vmselect",
+	})
+
+	testDeduplication(tc, sut, false)
+}
+
+func TestClusterDeduplication_deduplicationIsOn(t *testing.T) {
+	tc := at.NewTestCase(t)
+	defer tc.Stop()
+
+	sut := tc.MustStartCluster(&at.ClusterOptions{
```

Comment:
I think, it's worth to add replicationFactor=2 flag for vminsert.

Reply:
It should test deduplication performed by vmselect, when vmstorages have replicated data